pkg/nodeagent/files/files.go

```diff
@@ -0,0 +1,153 @@
+		}
+	} else if !os.IsNotExist(err) {
+		return err
+	}
+
+	if err := fs.MkdirAll(path.Dir(destination), permissions); err != nil {
+		return fmt.Errorf("destination directory %q could not be created", path.Dir(destination))
+	}
+
+	tempDir, err := fs.TempDir(v1alpha1.TempDir, "copy-image-")
+	if err != nil {
+		return fmt.Errorf("error creating temp directory: %w", err)
+	}
+
+	defer func() { runtime.HandleError(fs.Remove(tempDir)) }()
+
+	tmpFilePath := filepath.Join(tempDir, filepath.Base(destination))
+
+	if err := copyAndSync(fs, source, tmpFilePath); err != nil {
+		return err
+	}
+
+	if err := fs.Chmod(tmpFilePath, permissions); err != nil {
+		return err
+	}
+
+	return Move(fs, tmpFilePath, destination)
+}
+
+func copyAndSync(fs afero.Afero, source, destination string) error {
+	sourceFileStat, err := fs.Stat(source)
+	if err != nil {
+		return err
+	}
+	if !sourceFileStat.Mode().IsRegular() {
+		return fmt.Errorf("source %q is not a regular file", source)
+	}
+
+	if destinationFileStat, err := fs.Stat(destination); err == nil {
+		if !destinationFileStat.Mode().IsRegular() {
+			return fmt.Errorf("destination %q exists but is not a regular file", destination)
+		}
+	} else if !os.IsNotExist(err) {
+		return err
+	}
+
+	srcFile, err := fs.Open(source)
+	if err != nil {
+		return err
+	}
+	defer func() { runtime.HandleError(srcFile.Close()) }()
+
+	if err := fs.MkdirAll(path.Dir(destination), os.ModeDir); err != nil {
+		return fmt.Errorf("destination directory %q could not be created", path.Dir(destination))
+	}
+
+	dstFile, err := fs.OpenFile(destination, os.O_CREATE|os.O_RDWR|os.O_EXCL, 0700)
+	if err != nil {
+		return err
+	}
+	defer func() { runtime.HandleError(dstFile.Close()) }()
+
+	if _, err = io.Copy(dstFile, srcFile); err != nil {
+		return err
+	}
+
+	if err := fs.Chmod(dstFile.Name(), sourceFileStat.Mode()); err != nil {
+		return err
+	}
+
+	return dstFile.Sync()
+}
+
+// CrossDeviceModeOnly is exposed for testing cross-device moving mode.
+var CrossDeviceModeOnly = false
+
+// Move moves files on the same or across different devices.
+func Move(fs afero.Afero, source, destination string) error {
+	sourceFileStat, err := fs.Stat(source)
+	if err != nil {
+		return err
+	}
+	if !sourceFileStat.Mode().IsRegular() {
+		return fmt.Errorf("source %q is not a regular file", source)
+	}
+
+	if destinationFileStat, err := fs.Stat(destination); err == nil {
+		if !destinationFileStat.Mode().IsRegular() {
+			return fmt.Errorf("destination %q exists but is not a regular file", destination)
+		}
+	} else if !os.IsNotExist(err) {
+		return err
+	}
+
+	if CrossDeviceModeOnly {
+		return moveCrossDevice(fs, source, destination)
+	}
+
+	err = fs.Rename(source, destination)
+	if err != nil && strings.Contains(err.Error(), "cross-device link") {
+		return moveCrossDevice(fs, source, destination)
+	}
+	return err
+}
+
+func moveCrossDevice(fs afero.Afero, source, destination string) error {
+	tmpFilePath := destination + ".tmp"
```

Comment:
One last question. Shouldn't this delete the file at `tmpFilePath` first? `copyAndSync` could fail otherwise if there's already a file with that name.